### PR TITLE
Implement Public Allocator

### PR DESCRIPTION
**Pull request by adhusson**
### Summary
**tldr**
Main function is `reallocateTo(Withdrawal[] calldata withdrawals, MarketParams calldata supplyMarketParams)`, it is callable by anyone. `withdrawals` is a list of `(market,amount)` pairs. The function will try to withdraw the given `amount` from each market and then deposit the total withdrawn to `depositMarketParams`, subject to admin-settable constraints.

**Max flows**
Each market has a max inflow (`maxIn`) and a max outflow (`maxOut`). For every call to `reallocateTo`: 
* The markets in `withdrawals` increase their `maxIn` and decrease their `maxOut` by the amount withdrawn.
* The market `depositMarketParams` increases its `maxOut` and decreases its `maxIn` by the total withdrawn.

The owner can configure the max inflows/outflows.

**Supply caps**
Each market has a supply cap that works like MetaMorpho's supply caps. The owner can configure the caps.

**Fee**
The `PublicAllocator` has a fee (payable in ETH). The owner can configure the fee. The owner must call `transferFee` to take the accumulated fees out of the `PublicAllocator`.

**Checks**
The `PublicAllocator` will make the following checks when `reallocateTo` is called:
- No market's `maxIn` or `maxOut` is exceeded.
- No market's supply cap is exceeded.
- The fee has been paid.

### CODE REVIEW

File: src/libraries/UtilsLib.sol [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+library UtilsLib {
+    /// @dev Returns `x` safely cast to uint128.
+    function toUint128(uint256 x) internal pure returns (uint128) {
+        if (x > type(uint128).max) {
+            revert ErrorsLib.MaxUint128Exceeded();
+        }
+        return uint128(x);
+    }
+
+    // Returns min(x+y,type(uint128).max)
```
Comment (QGarchery) [inline src/libraries/UtilsLib.sol:20]:
```suggestion
    /// @dev Returns min(x+y,type(uint128).max).
```

File: src/libraries/UtilsLib.sol [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    function saturatingAdd(uint128 x, uint128 y) internal pure returns (uint128 z) {
+        assembly ("memory-safe") {
+            let sum := add(x, y)
+            let ceil := 0x00000000000000000000000000000000ffffffffffffffffffffffffffffffff
+            z := xor(sum, mul(xor(sum, ceil), lt(ceil, sum)))
+        }
+    }
```
Comment (QGarchery) [inline src/libraries/UtilsLib.sol:27]: We could add a test for it

File: src/libraries/EventsLib.sol [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+import {FlowConfig, SupplyConfig} from "../interfaces/IPublicAllocator.sol";
+
+/// @title EventsLib
+/// @author Morpho Labs
+/// @custom:contact security@morpho.org
+/// @notice Library exposing events.
+library EventsLib {
+    /// @notice Emitted when the public reallocation is triggered.
+    event PublicReallocate(address sender, uint256 fee);
```
Comment (MathisGD) [inline src/libraries/EventsLib.sol:12]: the fee cannot be something else actually
```suggestion
    event PublicReallocate(address sender);
```

File: src/libraries/ErrorsLib.sol [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    error InflowCapExceeded(Id id);
+
+    /// @notice Thrown when allocation from market `id` exceeds current max outflow
+    error OutflowCapExceeded(Id id);
+
+    /// @notice Thrown when flow configuration for market `id` has min flow > max flow
+    error InconsistentFlowConfig(Id id);
+
+    /// @notice Thrown when the reallocation fee given is wrong
+    error IncorrectFee(uint givenFee);
```
Comment (MathisGD) [inline src/libraries/ErrorsLib.sol:27]: I don't think that it's useful (you know how much you where paying no?)
```suggestion
    error IncorrectFee();
```

File: src/libraries/UtilsLib.sol [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+/// @author Morpho Labs
+/// @custom:contact security@morpho.org
+/// @notice Library exposing helpers.
+/// @dev Inspired by https://github.com/morpho-org/morpho-utils.
+library UtilsLib {
+    /// @dev Returns `x` safely cast to uint128.
+    function toUint128(uint256 x) internal pure returns (uint128) {
+        if (x > type(uint128).max) {
+            revert ErrorsLib.MaxUint128Exceeded();
+        }
```
Comment (MathisGD) [inline src/libraries/UtilsLib.sol:16]: 🤓 
```suggestion
        if (x > type(uint128).max) revert ErrorsLib.MaxUint128Exceeded();
```

File: src/libraries/UtilsLib.sol [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+pragma solidity ^0.8.0;
+
+import {ErrorsLib} from "../libraries/ErrorsLib.sol";
+
+/// @title UtilsLib
+/// @author Morpho Labs
+/// @custom:contact security@morpho.org
+/// @notice Library exposing helpers.
+/// @dev Inspired by https://github.com/morpho-org/morpho-utils.
+library UtilsLib {
```
Comment (MathisGD) [inline src/libraries/UtilsLib.sol:11]: note that we could import `toUint128` and `min` from Blue's Utils (but i'm not sure it's better)

File: src/PublicAllocator.sol [hunk 7]
```diff
@@ -1,25 +1,127 @@
...
+        MarketParams memory marketParams;
+        for (uint256 i = 0; i < allocations.length; ++i) {
+            marketParams = allocations[i].marketParams;
+            Id id = marketParams.id();
+            uint256 newAssets = MORPHO.expectedSupplyAssets(marketParams, address(VAULT));
+            if (newAssets >= assets[i]) {
+                if (newAssets > supplyCap[id]) revert ErrorsLib.PublicAllocatorSupplyCapExceeded(id);
+                uint128 inflow = (newAssets - assets[i]).toUint128();
+                flowCap[id].maxIn -= inflow;
+                flowCap[id].maxOut = (flowCap[id].maxOut).saturatingAdd(inflow);
```
Comment (MathisGD) [inline src/PublicAllocator.sol:83]: also possible
```suggestion
                flowCap[id].maxOut = uint128(min(flowCap[id].maxOut + inflow, type(uint128).max));
```

File: src/PublicAllocator.sol [hunk 8]
```diff
@@ -1,25 +1,127 @@
...
+import {EventsLib} from "./libraries/EventsLib.sol";
+import {FlowCap, FlowConfig, SupplyConfig, IPublicAllocatorStaticTyping} from "./interfaces/IPublicAllocator.sol";
+
+contract PublicAllocator is IPublicAllocatorStaticTyping {
+    using MorphoLib for IMorpho;
+    using MorphoBalancesLib for IMorpho;
+    using MarketParamsLib for MarketParams;
+    using SharesMathLib for uint256;
+    using UtilsLib for uint256;
+    using UtilsLib for uint128;
```
Comment (MathisGD) [inline src/PublicAllocator.sol:29]:
```suggestion
    using MorphoBalancesLib for IMorpho;
    using MarketParamsLib for MarketParams;
    using UtilsLib for uint256;
    using UtilsLib for uint128;
```

File: src/PublicAllocator.sol [hunk 9]
```diff
@@ -1,25 +1,127 @@
...
 
-contract PublicAllocator {
+import {MarketParamsLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/MarketParamsLib.sol";
+
+import {MorphoLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/periphery/MorphoLib.sol";
+
+import {MorphoBalancesLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/periphery/MorphoBalancesLib.sol";
+
+import {SharesMathLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/SharesMathLib.sol";
+
```
Comment (MathisGD) [inline src/PublicAllocator.sol:15]:
```suggestion
```

File: src/PublicAllocator.sol [hunk 10]
```diff
@@ -1,25 +1,127 @@
...
-import {IMetaMorpho, MarketAllocation} from "../lib/metamorpho/src/interfaces/IMetaMorpho.sol";
+import {
+    Id, IMorpho, IMetaMorpho, MarketAllocation, MarketParams
+} from "../lib/metamorpho/src/interfaces/IMetaMorpho.sol";
 
-contract PublicAllocator {
+import {MarketParamsLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/MarketParamsLib.sol";
+
+import {MorphoLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/periphery/MorphoLib.sol";
+
```
Comment (MathisGD) [inline src/PublicAllocator.sol:11]:
```suggestion
```

File: src/PublicAllocator.sol [hunk 11]
```diff
@@ -1,25 +1,127 @@
...
+        if (msg.value != fee) revert ErrorsLib.IncorrectFee(msg.value);
+
+        uint256[] memory assets = new uint256[](allocations.length);
+        for (uint256 i = 0; i < allocations.length; ++i) {
+            // Do not compute interest twice for every market
+            MORPHO.accrueInterest(allocations[i].marketParams);
+            assets[i] = MORPHO.expectedSupplyAssets(allocations[i].marketParams, address(VAULT));
+        }
+
+        VAULT.reallocate(allocations);
```
Comment (MathisGD) [inline src/PublicAllocator.sol:72]: nb: we would be very happy if MM was returning the changes applied to the markets

File: src/PublicAllocator.sol (hunk 7 above)
Comment (MathisGD) [inline src/PublicAllocator.sol:83]: if you put twice the same market in `allocations`, `maxIn` and `maxOut` are updated twice no? It seems that you can game the publicAllocator this way.

File: src/PublicAllocator.sol [hunk 12]
```diff
@@ -1,25 +1,127 @@
...
+            flowCap[flowCaps[i].id] = flowCaps[i].cap;
+        }
+
+        emit EventsLib.SetFlowCaps(flowCaps);
+    }
+
+    // Set supply cap. Public reallocation will not be able to increase supply if it ends above its cap.
+    // Doesn't revert if it doesn't change the storage at all
+    function setSupplyCaps(SupplyConfig[] calldata supplyCaps) external onlyOwner {
+        for (uint256 i = 0; i < supplyCaps.length; ++i) {
```
Comment (MathisGD) [inline src/PublicAllocator.sol:121]: I think that this is no longer an opti
```suggestion
        for (uint256 i = 0; i < supplyCaps.length; i++) {
```

File: src/PublicAllocator.sol [hunk 13]
```diff
@@ -1,25 +1,127 @@
...
+    function transferFee(address payable feeRecipient) external onlyOwner {
+        uint256 balance = address(this).balance;
+        feeRecipient.transfer(balance);
+        emit EventsLib.TransferFee(balance);
+    }
+
+    // Set flow cap
+    // Doesn't revert if it doesn't change the storage at all
+    function setFlowCaps(FlowConfig[] calldata flowCaps) external onlyOwner {
+        for (uint256 i = 0; i < flowCaps.length; ++i) {
```
Comment (MathisGD) [inline src/PublicAllocator.sol:111]:
```suggestion
        for (uint256 i = 0; i < flowCaps.length; i++) {
```

File: src/libraries/UtilsLib.sol (hunk 6 above)
Comment (adhusson) [reply src/libraries/UtilsLib.sol:11]: Had this at first but when I added `saturatingAdd` it seemed cleaner to have a small dedicated lib.

File: src/interfaces/IPublicAllocator.sol [hunk 14]
```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.21;
+
+import {IMetaMorpho, IMorpho, MarketAllocation, Id} from "../../lib/metamorpho/src/interfaces/IMetaMorpho.sol";
+
+struct FlowCap {
```
Comment (MathisGD) [inline src/interfaces/IPublicAllocator.sol:6]:
```suggestion
struct FlowCaps {
```

File: src/interfaces/IPublicAllocator.sol [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    /// @notice The maximum allowed inflow in a market
+    uint128 maxIn;
+    /// @notice The maximum allowed outflow in a market
+    uint128 maxOut;
+}
+
+struct FlowConfig {
+    Id id;
+    FlowCap cap;
```
Comment (MathisGD) [inline src/interfaces/IPublicAllocator.sol:15]:
```suggestion
    FlowCaps flowCaps;
```

File: src/interfaces/IPublicAllocator.sol [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+struct SupplyConfig {
+    Id id;
+    uint256 cap;
```
Comment (MathisGD) [inline src/interfaces/IPublicAllocator.sol:20]:
```suggestion
    uint256 supplyCap;
```

File: src/interfaces/IPublicAllocator.sol [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+import {IMetaMorpho, IMorpho, MarketAllocation, Id} from "../../lib/metamorpho/src/interfaces/IMetaMorpho.sol";
+
+struct FlowCap {
+    /// @notice The maximum allowed inflow in a market
+    uint128 maxIn;
+    /// @notice The maximum allowed outflow in a market
+    uint128 maxOut;
+}
+
+struct FlowConfig {
```
Comment (MathisGD) [inline src/interfaces/IPublicAllocator.sol:13]:
```suggestion
struct FlowCapsConfig {
```

File: src/interfaces/IPublicAllocator.sol [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+    Id id;
+    FlowCap cap;
+}
+
+struct SupplyConfig {
```
Comment (MathisGD) [inline src/interfaces/IPublicAllocator.sol:18]:
```suggestion
struct SupplyCapConfig {
```

File: src/PublicAllocator.sol (hunk 7 above)
Comment (QGarchery) [reply src/PublicAllocator.sol:100]: Yes, this is the spec of `saturatingAdd`

Comment (MathisGD) [reply src/PublicAllocator.sol:100]: yes actually 😄

File: src/libraries/EventsLib.sol (hunk 3 above)
Comment (adhusson) [reply src/libraries/EventsLib.sol]: 34cac9fb9e4994ae651490c4c09a8973e1051711

File: src/libraries/ErrorsLib.sol (hunk 4 above)
Comment (adhusson) [reply src/libraries/ErrorsLib.sol]: 34cac9fb9e4994ae651490c4c09a8973e1051711

File: src/PublicAllocator.sol (hunk 7 above)
Comment (adhusson) [reply src/PublicAllocator.sol:100]: Damn yes, you can make the public allocator overestimate the flows.

File: src/PublicAllocator.sol (hunk 12 above)
Comment (adhusson) [reply src/PublicAllocator.sol:136]: d7b76268069f5c04da7bef8c93c76fbca278d88b

File: src/PublicAllocator.sol (hunk 13 above)
Comment (adhusson) [reply src/PublicAllocator.sol:127]: d7b76268069f5c04da7bef8c93c76fbca278d88b

File: src/interfaces/IPublicAllocator.sol (hunk 17 above)
Comment (adhusson) [reply src/interfaces/IPublicAllocator.sol:13]: FlowCapsConfig.flowCaps vs FlowCapsConfig.caps vs FlowConfig.cap, at least the first seems too long.

Comment (adhusson) [reply src/interfaces/IPublicAllocator.sol:13]: What about

- `FlowConfig { Id id; FlowRange range }`
- `SupplyConfig { Id id; uint cap }`

(or for me `FlowCap` being 2 fields is fine)

File: src/PublicAllocator.sol [hunk 19]
```diff
@@ -1,25 +1,126 @@
...
     /// CONSTRUCTOR ///
 
-    constructor(address newOwner, IMetaMorpho newVault) {
-        owner = newOwner;
-        vault = newVault;
+    constructor(address owner, address vault) Ownable(owner) {
+        if (vault == address(0)) {
+            revert ErrorsLib.ZeroAddress();
+        }
+        VAULT = IMetaMorpho(vault);
```
[the remark this answers is not in the document]
Comment (StErMi) [reply src/PublicAllocator.sol:70]: > I'm not seeing what bad scenario it helps prevent?

It depends mostly on how you are using the factory's vault for and which are the requirements for these public allocators. 

If by specs you want to allow public allocators to interact with "official" vaults generated via the `MorphoFactory`, you need to add that check; otherwise you can leave it open to any vault.

I hope I was clear on what I meant :D

File: src/libraries/UtilsLib.sol (hunk 6 above)
Comment (MathisGD) [reply src/libraries/UtilsLib.sol:11]: just realized that elsewhere we would have called it "Uint128MaxCeilAdd" ("ZeroFloorSub")

File: src/PublicAllocator.sol (hunk 19 above)
[the remark this answers is not in the document]
Comment (adhusson) [reply src/PublicAllocator.sol:70]: I followed your suggestion in the `PublicAllocatorFactory` (41f205662d74590e436f1e4b548380ce7e77b970). That gives two tracks:
- With guardrails: you do everything through the factories, and if `publicAllocatorFactory.isPublicAllocator(account)`, then you know you're using the right contracts all the way down.
- Without guardrails: you can deploy your own `PublicAllocator` and connect it to your own `MetaMorpho`.
